### PR TITLE
feat(hmr): Add client and server hmr for react

### PR DESCRIPTION
**Pull request by nicolassenechal**
## Commit Message For Review
Add React hmr to sku

RFC URL:
https://github.com/seek-oss/sku/issues/62

REASON FOR CHANGE:
This PR aims at facilitating the use of sku for react projects by enabling hmr.

EXAMPLE USAGE:
- You can now provide, as part of your sku config, an array with two port numbers to your build when using SSR rendering. The first one will be used to start the dev server used for client-side hmr. The second one is used for the server-side hmr. You access your application using the second port. 

```js
module.exports = {
  entry: {
    client: 'src/client.js',
    server: 'src/server/server.js'
  },
  public: 'src/public',
  target: 'dist',
  port: [3000, 3300],
  compilePackages: ['seek-asia-style-guide']
}
```

- There is no change if you use static rendering, though this is untested. Static rendering with hmr on the client-side should work

- To enable hmr in your ssr application, you now need to load the client from the client server as those are different
```js
<script type="text/javascript" src="http://localhost:3000/main.js"></script>
```

- To enable hmr client-side, you need to add an AppContainer and define the behaviour whenever a rebuild happens as described in here: https://github.com/gaearon/react-hot-loader/tree/master/docs#migration-to-30
```js
import React from 'react';
import { render } from 'react-dom';
import App from './App/App';
import { AppContainer } from 'react-hot-loader';

const appElement = document.getElementById('app');
render(
  <AppContainer>
      <App />
  </AppContainer>,
  appElement
);

if (module.hot) {
  module.hot.accept('./App/App', () => {
    render(
      <AppContainer>
          <App />
      </AppContainer>,
      appElement
    );
  });
}
```

NOTES:
- Setup tested only against our team project which is not a mono-repo and uses server-side rendering. Happy to validate against other projects, but this emphasises the need for a project we want to test changes against
- The webpack config build function does not test anymore against the environment, but against the script being run. Happy to get rid of the production tests across the whole file or any alternative solution.
- If you use SSR, the dist folder needs to be deleted before starting the server: `rm -rf ./dist && sku start`

### CODE REVIEW

File: config/webpack/webpack.config.js [hunk 1]
```diff
@@ -124,31 +127,45 @@ const buildWebpackConfigs = builds.map(
         return JSON.stringify(valueForEnv);
       })
       .set('SKU_ENV', JSON.stringify(args.env))
-      .set('PORT', JSON.stringify(port))
+      .set('PORT', JSON.stringify(port.length > 1 ? port[1] : ''))
```
Comment (mengtzu) [inline config/webpack/webpack.config.js]: Need to be PORT or can it be SKU_PORT or similar?

File: config/server/server.js [hunk 2]
```diff
@@ -19,6 +19,10 @@ if (middleware) {
 }
 app.get('*', renderCallback);
 
-app.listen(port, () => {
-  console.log(`App started on port ${port}`);
-});
+if (env === 'production') {
+  app.listen(port, () => {
```
Comment (mengtzu) [inline config/server/server.js]: This port also gets listened to in server/index right?  Why is this necessary?

**Comment by mengtzu** [comment]
Apart from above comments, looking good.  It does complicate an already complicated set of config but I don't see how to escape that.

File: config/webpack/webpack.config.js (hunk 1 above)
Comment (nicolassenechal) [reply config/webpack/webpack.config.js]: Good point, worth updating

File: config/server/server.js (hunk 2 above)
Comment (nicolassenechal) [reply config/server/server.js]: server/index is only for dev. This is for production

Comment (nicolassenechal) [reply config/server/server.js]: Which implied you found a bug :|

**Comment by mattsjones** [comment]
So I just tried this against profile which is already running `sku`. I'm getting a `Cannot GET /` error. Are there any necessary changes I need to make?

**Comment by nicolassenechal** [comment]
Previous issue should be fixed

**Comment by nicolassenechal** [comment]
Changed port API. Ready for review

**Comment by markdalgleish** [comment]
Even though there's some duplication now, I think this is so much better, and we can easily extract the common code into separate files where appropriate 👍

**Comment by markdalgleish** [comment]
Don't forget to update the description, release notes etc. to feature the new API.
